stdlib: Reworking AST queries

skberkeley opened the pull request:
Initial reworking of AST queries to try to make authoring codemods a bit more ergonomic. APIs are still in flux, so open to suggestions!

Follow up work still needed:
- Reworking `query:replace` so programmers don't need to pass replacement maps around

skberkeley committed:
query updates

skberkeley committed:
query tests

skberkeley committed:
things got luacased

skberkeley committed:
less annotations needed

hgoldstein commented on examples/query.luau:
```diff
@@ -0,0 +1,25 @@
+		return firstArg.tag == "indexname" and firstArg.index.text == "OldComponent"
+	end)
+
+-- Can we improve the type solver here so we don't need to add this annotation?
```
Which annotation are you talking about?

skberkeley replied:
On the line right after:
<img width="1054" height="318" alt="Image" src="https://github.com/user-attachments/assets/6c9e0af1-645c-49e0-b4b0-b6bc092a31aa" />

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,163 @@
-		if result then
-			table.insert(nodes, result)
+type replacement = string | node | { delta }
+type replacements = { [node]: replacement }
```
Should types here be capitalized?

skberkeley replied:
I think the latest in lute naming conventions is to luacase types too (can you confirm @aatxe ?)

Vighnesh-V replied:
exported types should be luacase as far as I can tell, but this is alright if it's not being exported.

skberkeley replied:
it'll be exported later, so i'll keep it as is!

hgoldstein reviewed (commented):
Would you be able to add an end-to-end test? It feels weird to see all of this infrastructure without being able to _actually_ run a codemod.

hgoldstein commented on definitions/luau.luau:
```diff
@@ -211,7 +212,7 @@ export type AstExprIfElse = {
 }
 
-export type AstExpr = { kind: "expr" } & (
+export type AstExpr = { kind: "expr"? } & (
```
Why is this optional?

skberkeley replied:
This not being optional means checks like `AstExprConstantNil <: AstExpr` doesn't subtype, but I think the better thing to do might be to add a `kind` field to each variant too

hgoldstein replied:
I think so, this also means that:
```
if e.kind then
end
```
... would exclude those fields, which seems incorrect.

skberkeley committed:
allow replacements

hgoldstein commented on tests/syntax/query.test.luau:
```diff
@@ -0,0 +1,166 @@
+			createNumberExpr(0),
+			createNumberExpr(1),
+			createNumberExpr(2),
+		})
```
Why not just parse this as normal Luau code?

hgoldstein replied:
That's not what I meant. Why isn't this:
```luau
local testFile = [[
    local numbers = {0, 1, 2, 3, 4}
]]
```

Ditto for the rest of this file.

skberkeley replied:
oh i see what you mean
i wanted to be able to unit test the other functions in the query library without using `select`. i could parse `local numbers = {0, 1, 2, 3, 4}` as a whole and wrangle the `AstExprConstantNumbers` out of the AST, but this seems more concise for the same effect?

hgoldstein replied:
> i wanted to be able to unit test the other functions in the query library without using `select`

That doesn't seem like a meaningful test to me: it's just visual noise.

wmccrthy commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,163 @@
+		visitTypePackGeneric = visit,
+		visitTypePackVariadic = visit,
+
+		visitToken = function(n)
```
I think we should use `visit` here too. I've found it's often useful to query specifically for tokens, since it's a straightforward way to search for and replace exact text

skberkeley replied:
I originally did use `visit` here, but the trouble is that we have `AstExprs` like `AstExprConstantNumber` which are technically tokens too, which means that the selection function gets called twice on them, so we end up overcounting in a sense. What do you think about building out a `selectToken` in the future which only selects tokens?

wmccrthy replied:
Ah I gotcha, that makes sense. I think 'selectToken' would be a very nice addition then.

wmccrthy commented on lute/std/libs/syntax/utils.luau:
```diff
@@ -0,0 +1,229 @@
+function utils.isTypePack(n: luau.AstNode): luau.AstTypePack?
+	return if n.kind == "typepack" then n else nil
+end
+
```
same vein as my other comment, but maybe worth adding `isToken`
```suggestion

function utils.isToken(n: luau.AstNode): luau.Token?
	return if n.istoken then n else nil
end

```

wmccrthy commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,163 @@
+}
+
+local queryLib = {}
+
```
spoken abt offline, but `forEach` would be nice
```suggestion

function queryLib.forEach<T>(self: query<T>, callback: (T) -> ())
	for _, node in self.nodes do
		callback(node)
	end
end

```

skberkeley committed:
non-optional kinds

skberkeley committed:
map can flatten too

skberkeley committed:
transformer example should have local function

skberkeley committed:
ast reserialization + tests

skberkeley committed:
parse instead of mocking number exprs

skberkeley committed:
istoken util

skberkeley committed:
foreach

skberkeley committed:
call printInterpString

skberkeley committed:
printer tests

skberkeley commented:
Added AST replacement reserialization! I've added unit tests for reserialization stuff, as well as a basic sanity test for a codemod using the new APIs and @wmccrthy has volunteered to write a more complete suite of E2E tests

skberkeley committed:
Merge branch 'primary' into skberkeley/ast_queries

skberkeley committed:
table library got renamed

hgoldstein commented on tests/std/syntax/printer.test.luau:
```diff
@@ -0,0 +1,236 @@
+
+		local result = printer.printstatement(ast, replacements)
+
+		local expected = "            local name = 1 + 2" -- TODO: preserve trivia so trailing newline is preserved
```
That's not a `TODO`, that's an important part of the contract for dealing with CSTs. Also what happened to the newline at the beginning of the string?

skberkeley replied:
I agree, i'm definitely not trying to claim that this PR makes ast querying/reserialization feature complete, but in the interest of not letting this PR grow indefinitely, I was planning on handling trivia reserialization in a follow up

> Also what happened to the newline at the beginning of the string?

There never was a newline:

```luau
local source = [[
print(`Hello {name}`)
]]

print(string.byte(source, 1))
```
```
% lute scratch.luau 
112
```

hgoldstein commented on definitions/luau.luau:
```diff
@@ -104,13 +99,15 @@ export type AstExprIndexName = {
 	closebrackets: Token<"]">,
 }
 
+-- I don't like that we have this type; a lot of its data should live in the parents, which should then have `AstStatBlock` as its body
```
If this is how the AST is structured, then we match it, especially for a CST.

skberkeley replied:
It's not how the AST or CST are structured in the C++ implementation, do you think it's worth refactoring sooner rather than later?

hgoldstein replied:
Why is it like this then? We should probably refactor it _soon_ but out of scope for this PR.

skberkeley replied:
I hate to be that guy, but these type definitions were written before I started working on this stuff

hgoldstein commented on examples/query_transformer.luau:
```diff
@@ -0,0 +1,24 @@
+		return `math.isnan({(bin.lhsoperand :: luau.AstExprLocal).token.text})`
+	end)
+
+	return printer.printfile(parsedResult, replacements)
```
Why is this not one long expression like:
```
query.select(parsedResult.root, syntax_utils.isExprBinary)
    :filter(function(bin)
        return bin.operator.text == "~="
            and bin.lhsoperand.tag == "local"
            and bin.rhsoperand.tag == "local"
            and bin.lhsoperand.token.text == bin.rhsoperand.token.text
    end)
    :replace(function(bin)
        return `math.isnan({(bin.lhsoperand :: luau.AstExprLocal).token.text})`
    end)
    :print()
```

skberkeley replied:
oops yeah, i can make that change! i think i was trying to debug autocomplete at the time

hgoldstein replied:
Did you update this file?

hgoldstein commented on lute/std/libs/syntax/printer.luau:
```diff
@@ -2,157 +2,292 @@
+			error("TODO: Add visit function to visit library")
+		else
+			error(`Unknown node kind in replacement: {replacement.kind}`)
+		end
```
This feels bizarre: why do we need to match on individual node kinds here? Why do we have random holes?

skberkeley replied:
You're right, I'll add a general `visit` function to the visitor library, it shouldn't be on consumers to check the node kind

hgoldstein commented on lute/std/libs/syntax/printer.luau:
```diff
@@ -2,157 +2,292 @@
+			if delta.tag == "append" then
+				self:printReplacement(node, delta.content)
+			end
+		end
```
Can we do something more constrained here? It feels weird if we have a single node and there are _multiple_ replacements. How about the rule for now is that you get one `delta`?

skberkeley replied:
Sure, I suppose codemod authors can always just specify longer strings in their deltas, so the expressivity isn't reduced

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,173 @@
+	elseif nodeKind == "token" then
+		visitor.visittoken(ast :: luau.Token, selectVisitor)
+	else
+		error("Unsupported node kind: " .. tostring(nodeKind))
```
Could this be exhaustive?

skberkeley replied:
Yuh, it should be once I add a general visit function to the visitor library

hgoldstein commented on lute/std/libs/syntax/types.luau:
```diff
@@ -0,0 +1,15 @@
+export type delta =
+	| { tag: "append", content: string | node }
+	| { tag: "prepend", content: string | node }
+	| { tag: "replace", content: string | node }
```
The content for deltas should _always_ be `string`.

skberkeley replied:
is there no world where codemod writers might benefit from being able to construct an AST node object from scratch and using it as a delta content?

hgoldstein replied:
_Potentially_, but deltas are meant to be lower level components.

hgoldstein reviewed (changes requested):
See comments.

Vighnesh-V commented on examples/query.luau:
```diff
@@ -0,0 +1,25 @@
+
+local ast = luau.parseexpr("require(OldComponent)")
+
+local p = query
+	.select(ast, utils.isExprCall)
+	:filter(function(call)
+		local calledFunc = call.func
+		return calledFunc.tag == "global" and calledFunc.name.text == "require"
+	end)
+	:filter(function(call)
+		local firstArg = call.arguments[1].node
+		return firstArg.tag == "indexname" and firstArg.index.text == "OldComponent"
+	end)
```
this is pretty sick.

Vighnesh-V reviewed (approved):
This is dope - it looks good to me

skberkeley committed:
add a general visit function

skberkeley committed:
only allow single deltas

skberkeley committed:
hunter comment

skberkeley committed:
Merge branch 'primary' into skberkeley/ast_queries

hgoldstein commented on lute/std/libs/syntax/printer.luau:
```diff
@@ -2,157 +2,262 @@
+	if typeof(replacement) == "string" then
+		self:write(replacement)
+	elseif replacement.kind ~= nil then -- LUAUFIX: type solver should know replacement might have a kind field
+		-- TODO: preserve trivia
```
Is this still `TODO`?

skberkeley committed:
cut deltas for now

skberkeley committed:
query test change

skberkeley committed:
add print method to replacements

skberkeley committed:
move query tests

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,168 @@
+	end
+end
+
+function queryLib.select<T>(ast: node, fn: (node) -> (T | { T })?): query<T>
```
Can we not have select flat-map? Select should just be an entrypoint, we can have map and flat-map as separate functions.

skberkeley replied:
Done! Also got rid of it from `map`

hgoldstein commented on examples/query_transformer.luau:
```diff
@@ -0,0 +1,23 @@
+	local parsedResult = parser.parsefile(ctx.source)
+
+	return query
+		.select(parsedResult.root, syntax_utils.isExprBinary)
```
Can this be named something different for now? We talked about this offline, that `select` really ought to be a method on `query`, but the new solver doesn't quite do the right thing here yet. Maybe name this `query.selectFromRoot`

skberkeley replied:
done!

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,168 @@
 
-	visitor.visitblock(block, finder)
+	local selectVisitor: visitor.Visitor = {
+		visitBlock = visit,
```
Maybe we're missing a building block here: kind of annoying that you have to make this literal with a ton of methods that almost all do the same thing?

skberkeley replied:
yeah i'll add a `visitor.createVisitor(defaultVisit: (AstNode -> boolean)?)` in a followup

hgoldstein commented on tests/std/syntax/query.test.luau:
```diff
@@ -0,0 +1,148 @@
+	suite:case("select_nums", function(assert)
+		local ast = luau.parse("print(1 + 2)")
+		local queryResult: query.query<luau.AstExprConstantNumber> = query.select(
+			ast.root :: luau.AstNode,
```
why do we need this type cast?

skberkeley replied:
Solver complains that AstStatBlock doesn't have location? field, so it doesn't subtype against AstNode

hgoldstein commented on tests/std/syntax/query.test.luau:
```diff
@@ -0,0 +1,148 @@
+		local num: luau.AstExpr = queryResult.nodes[1]
+		assert.eq(num.kind, "expr")
+		assert.eq(num.tag, "number")
+		assert.eq((num :: luau.AstExprConstantNumber).value, 1)
```
What about this cast?

skberkeley replied:
the `assert.eq` calls above don't refine `num` to `AstExprConstantNumber`

skberkeley committed:
no more flat mapping

skberkeley committed:
add luaufix comment

skberkeley committed:
Merge branch 'primary' into skberkeley/ast_queries

skberkeley committed:
update query transformer example

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,159 @@
+		map = self.map,
+		forEach = self.forEach,
+		__root = self.__root,
+	}
```
This is to get around the recursive type restriction yeah?

skberkeley replied:
Actually I think this can just be in-place, since the recursive type restriction doesn't impact the runtime

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,159 @@
+	nodes: { T },
+	filter: (self: query<T>, pred: (T) -> boolean) -> query<T>,
+	replace: (self: query<T>, repl: (T) -> types.replacement?) -> types.replacements,
+	map: <U>(self: query<T>, fn: (T) -> (U | { U })?) -> query<U>, -- recursion violation reported
```
This signature is wrong now I think

skberkeley replied:
nice catch

hgoldstein reviewed (changes requested):
Just the signature being out of date

skberkeley committed:
map can be in place, update signature

skberkeley committed:
Merge branch 'primary' into skberkeley/ast_queries

hgoldstein commented on lute/std/libs/syntax/query.luau:
```diff
@@ -3,28 +3,152 @@
+	map: <U>(self: query<T>, fn: (T) -> U?) -> query<U>, -- recursion violation reported
+	forEach: (self: query<T>, callback: (T) -> ()) -> (),
+	__root: node,
+} -- TODO: Add a select: (self: query<T>, fn: (node) -> U?) -> query<U> method and a flatMap: <U>(self: query<T>, fn: (T) -> { U }?) -> query<U> method
```
nit: `flatMap` is `(T) -> { U }`: the `nil` is redundant given that `nil` is the same as `{}`

skberkeley replied:
oh tru

skberkeley committed:
nit

skberkeley committed:
Merge branch 'primary' into skberkeley/ast_queries

wmccrthy commented on examples/query_transformer.luau:
```diff
@@ -0,0 +1,20 @@
+		:replace(function(bin)
+			return `math.isnan({(bin.lhsoperand :: luau.AstExprLocal).token.text})`
+		end)
+		:print()
```
does this work? doesn't look like there is a print method on the `query` class

skberkeley replied:
yup, it does! the `replacements` type returned by `replace()` has a `print` method

wmccrthy replied:
oh nice, missed that 🔥
